### PR TITLE
improve NewScArray fastpath codegen

### PR DESCRIPTION
**Pull request by sigatrev**
I noticed some wasteful codegen in the fast path for NewScArray while investigating an unrelated issue.

The int array case on x64 and x86 was using constant encoding on the MissingValue value, which is not needed.

On ARM and ARM64, every store in all array types was being done as a separate LDIMM, which expand to 2 to 4 instructions each.

This commit changes the behavior on intel to not use constant encoding for int missing values, and on arm to load the value one time into a register, and then store the register as needed.

```
// current x64 NewScArray fast path for NativeIntArray.
    s116(rdx).u32   =  MOV            -2147483646 (0xFFFFFFFF80000002) [encoded: 0x13E6A855].u32
    s116(rdx).i32   =  XOR            s116(rdx).i32, <cookie> (value: 0xFFFFFFFF93E6A857).i32
    [s100(rcx).u64+24].i32 = MOV      s116(rdx).i32
    s117(rdx).u32   =  MOV            -2147483646 (0xFFFFFFFF80000002) [encoded: 0xFFFFFFFFB65A2516].u32
    s117(rdx).i32   =  XOR            s117(rdx).i32, <cookie> (value: 0x365A2514).i32
    [s100(rcx).u64+28].i32 = MOV      s117(rdx).i32
    s118(rdx).u32   =  MOV            -2147483646 (0xFFFFFFFF80000002) [encoded: 0xFFFFFFFF959E394F].u32
    s118(rdx).i32   =  XOR            s118(rdx).i32, <cookie> (value: 0x159E394D).i32
    [s100(rcx).u64+32].i32 = MOV      s118(rdx).i32
    s119(rdx).u32   =  MOV            -2147483646 (0xFFFFFFFF80000002) [encoded: 0x6217D90D].u32
    s119(rdx).i32   =  XOR            s119(rdx).i32, <cookie> (value: 0xFFFFFFFFE217D90F).i32
    [s100(rcx).u64+36].i32 = MOV      s119(rdx).i32
    s120(rdx).u32   =  MOV            -2147483646 (0xFFFFFFFF80000002) [encoded: 0xFFFFFFFF8266FF08].u32
    s120(rdx).i32   =  XOR            s120(rdx).i32, <cookie> (value: 0x266FF0A).i32
    [s100(rcx).u64+40].i32 = MOV      s120(rdx).i32
    s121(rdx).u32   =  MOV            -2147483646 (0xFFFFFFFF80000002) [encoded: 0xFFFFFFFFDCB51B11].u32
    s121(rdx).i32   =  XOR            s121(rdx).i32, <cookie> (value: 0x5CB51B13).i32
    [s100(rcx).u64+44].i32 = MOV      s121(rdx).i32

// new x64 NewScArray fast path for NativeIntArray.
    [s100.u64+24].i32 = MOV           -2147483646 (0xFFFFFFFF80000002).i32    #
    [s100.u64+28].i32 = MOV           -2147483646 (0xFFFFFFFF80000002).i32    #
    [s100.u64+32].i32 = MOV           -2147483646 (0xFFFFFFFF80000002).i32    #
    [s100.u64+36].i32 = MOV           -2147483646 (0xFFFFFFFF80000002).i32    #
    [s100.u64+40].i32 = MOV           -2147483646 (0xFFFFFFFF80000002).i32    #
    [s100.u64+44].i32 = MOV           -2147483646 (0xFFFFFFFF80000002).i32    #

// current ARM64 NewScArray fast path for NativeFloatArray (worst of the three types)
    s153(r1).u64    =  MOVZ           2 (0x2).i16, 0 (0x0).u8
    s153(r1).u64    =  MOVK           32768 (0x8000).i16, 16 (0x10).u8
    s153(r1).u64    =  MOVK           2 (0x2).i16, 32 (0x20).u8
    s153(r1).u64    =  MOVK           32768 (0x8000).i16, 48 (0x30).u8
    [s140(r0).u64+24].u64 = STR       s153(r1).u64
    s154(r1).u64    =  MOVZ           2 (0x2).i16, 0 (0x0).u8
    s154(r1).u64    =  MOVK           32768 (0x8000).i16, 16 (0x10).u8
    s154(r1).u64    =  MOVK           2 (0x2).i16, 32 (0x20).u8
    s154(r1).u64    =  MOVK           32768 (0x8000).i16, 48 (0x30).u8
    [s140(r0).u64+32].u64 = STR       s154(r1).u64
    s155(r1).u64    =  MOVZ           2 (0x2).i16, 0 (0x0).u8
    s155(r1).u64    =  MOVK           32768 (0x8000).i16, 16 (0x10).u8
    s155(r1).u64    =  MOVK           2 (0x2).i16, 32 (0x20).u8
    s155(r1).u64    =  MOVK           32768 (0x8000).i16, 48 (0x30).u8
    [s140(r0).u64+40].u64 = STR       s155(r1).u64
    s156(r1).u64    =  MOVZ           2 (0x2).i16, 0 (0x0).u8
    s156(r1).u64    =  MOVK           32768 (0x8000).i16, 16 (0x10).u8
    s156(r1).u64    =  MOVK           2 (0x2).i16, 32 (0x20).u8
    s156(r1).u64    =  MOVK           32768 (0x8000).i16, 48 (0x30).u8
    [s140(r0).u64+48].u64 = STR       s156(r1).u64
    s157(r1).u64    =  MOVZ           2 (0x2).i16, 0 (0x0).u8
    s157(r1).u64    =  MOVK           32768 (0x8000).i16, 16 (0x10).u8
    s157(r1).u64    =  MOVK           2 (0x2).i16, 32 (0x20).u8
    s157(r1).u64    =  MOVK           32768 (0x8000).i16, 48 (0x30).u8
    [s140(r0).u64+56].u64 = STR       s157(r1).u64

// new ARM64 NewScArray fast path for NativeFloatArray
    s150(r1).var    =  MOVZ           2 (0x2).i16, 0 (0x0).u8
    s150(r1).var    =  MOVK           32768 (0x8000).i16, 16 (0x10).u8
    s150(r1).var    =  MOVK           2 (0x2).i16, 32 (0x20).u8
    s150(r1).var    =  MOVK           32768 (0x8000).i16, 48 (0x30).u8
    [s137(r0).u64+24].var = STR       s150(r1).var
    [s137(r0).u64+32].var = STR       s150(r1).var
    [s137(r0).u64+40].var = STR       s150(r1).var
    [s137(r0).u64+48].var = STR       s150(r1).var
    [s137(r0).u64+56].var = STR       s150(r1).var

```

### CODE REVIEW

File: lib/Backend/Lower.cpp [hunk 1]
```diff
@@ -3745,10 +3755,12 @@ Lowerer::GenerateProfiledNewScArrayFastPath(IR::Instr *instr, Js::ArrayCallSiteI
         const IR::AutoReuseOpnd autoReuseHeadOpnd(headOpnd, func);
 
         GenerateMemInit(dstOpnd, Js::JavascriptNativeIntArray::GetOffsetOfWeakFuncRef(), IR::AddrOpnd::New(weakFuncRef, IR::AddrOpndKindDynamicFunctionBodyWeakRef, m_func), instr, isZeroed);
+
+        IR::Opnd * missingItemOpnd = getMissingItemOpnd(TyInt32);
         for (; i < size; i++)
         {
             GenerateMemInit(headOpnd, sizeof(Js::SparseArraySegmentBase) + i * sizeof(int32),
-                Js::JavascriptNativeIntArray::MissingItem, instr, isZeroed);
+                missingItemOpnd, instr, isZeroed);
```
Comment (Penguinwizzard) [inline lib/Backend/Lower.cpp]: Do we not need an autoReuseMissingItemOpnd here?

File: lib/Backend/Lower.cpp [hunk 2]
```diff
@@ -3784,11 +3800,15 @@ Lowerer::GenerateProfiledNewScArrayFastPath(IR::Instr *instr, Js::ArrayCallSiteI
...
+        const IR::AutoReuseOpnd autoReuseMissingItemOpnd(missingItemOpnd, func);
+
         for (; i < size; i++)
         {
             GenerateMemInit(
                 headOpnd, offsetStart + i * sizeof(Js::Var),
-                IR::AddrOpnd::New(Js::JavascriptArray::MissingItem, IR::AddrOpndKindConstantAddress, m_func, true),
+                missingItemOpnd,
                 instr, isZeroed);
         }
```
Comment (Penguinwizzard) [inline lib/Backend/Lower.cpp]: nit: These three blocks look extremely similar; is there potentially a benefit to patterning them out to three calls to a single lambda?

File: lib/Backend/Lower.cpp (hunk 1 above)
Comment (sigatrev) [reply lib/Backend/Lower.cpp]: I added the others in response to failures. It seemed like it wasn't needed here, but it wouldn't hurt to have. Either way, it's included in the refactor in my commit.

File: lib/Backend/Lower.cpp (hunk 2 above)
Comment (sigatrev) [reply lib/Backend/Lower.cpp]: Yeah, that's a good idea. done.

**Pull request merged**